### ISSUE TITLE
iterator: remove unreachable!()

### ISSUE THREAD

**Issue by rkuris** [ava-labs/firewood#349]
This is a foot gun if anyone changes the code. It would be better to have 
```rust
let (found_key, value) = match std::mem::take(&mut self.key_state) {
```
at the top of the `poll_next` method.

_Originally posted by @richardpringle in https://github.com/ava-labs/firewood/pull/346#discussion_r1393095817_

**Comment by rkuris** [comment]
It's possible if you compute the key and value as part of the outer loop. This is tricky due to the borrow checker.

**Comment by richardpringle** [comment]
Solved in #430
